### DISCUSSION TITLE
Trigger workflow from organization webhook

### DISCUSSION THREAD

**Discussion by mbhave** [community/community#50955] [Actions]
### Select Topic Area

Question

### Body

I'd like to be able to trigger a Github Actions workflow in one repo based on changes to any repository in another organization. It seems like I should be able to use `repository_dispatch` combined with an organization level webhook. However, since I can't customize the payload for the webhook, I can't add an `event_type` to it which would be needed for the workflow. The webhook also seems to be getting a `403`.

Is there a way for me to trigger a GH actions workflow based on changes to repositories in an organization?

**Comment by Anmol-Baranwal** [comment]
However, you should still be able to trigger a `repository_dispatch` event without an `event_type` by using the default event type.

Regarding the 403 error you're getting, it could be caused by insufficient permissions. Make sure that the user or token you're using has the appropriate permissions to trigger workflows and create webhooks. You may also want to check if the organization has any restrictions on creating webhooks.


Here's how you can do it:

1. Create a new workflow file in the repository where you want to trigger the workflow. The file should be located in the `.github/workflows/` directory and should contain the steps you want to run when the workflow is triggered.

2. Add a `repository_dispatch` event to the workflow file. This event will trigger the workflow when a webhook is sent to the repository. Here's an example:

```yaml
on:
  repository_dispatch:
    types: [my-event]
```

Note that `my-event` is a custom event type that you can define. You'll need to remember this event type when creating the webhook.

3. Create a webhook on the organization level. Go to the organization's settings, select "Webhooks", and click "Add webhook". Set the payload URL to the URL for the repository's dispatch endpoint, which should be `https://api.github.com/repos/<OWNER>/<REPO>/dispatches` (replace `<OWNER>` and `<REPO>` with the appropriate values). Set the content type to "application/json". Finally, set the "Which events would you like to trigger this webhook?" option to "Let me select individual events" and select the "Repositories" event.

4. When creating the webhook, add a custom header with the name "X-GitHub-Event" and the value "my-event", where "my-event" is the custom event type you defined in step 2.

5. When you make changes to any repository in the organization, the webhook will be triggered and a POST request will be sent to the repository's dispatch endpoint with the custom event type in the payload. This will trigger the workflow in the repository you set up in step 1.

> Do tell me, if this doesn't resolve your issue.

**Reply by mbhave** [reply]
Thanks for your reply @Anmol-Baranwal. I can't seem to find how to add the custom header on the org webhook to include `my-event`. There isn't any option to add a custom header.

**Reply by mbhave** [reply]
Also, the 403 is not while creating the webhook. It's the webhook delivery that results in a 403 because there isn't way to specify an Authorization header which is needed for repository dispatches in Github Actions.

**Reply by Anmol-Baranwal** [reply]
> I apologize for the delay, it seems our conversation got lost amidst other discussions. I was reviewing my past discussions today and came across this one. I recognize that I was not being attentive, and I am sorry for any inconvenience or frustration that my behavior may have caused.

You can add the header in the webhook payload itself.

Here's an example of how you can add the custom header in the webhook payload:

```json
{
  "action": "created",
  "repository": {
    "full_name": "<OWNER>/<REPO>"
  },
  "organization": {
    "login": "<ORGANIZATION>"
  },
  "client_payload": {
    "event_type": "my-event"
  },
  "headers": {
    "Authorization": "Bearer <TOKEN>",
    "X-GitHub-Event": "my-event"
  }
}
```

In this example, the custom header "X-GitHub-Event" is added to the payload with the value "my-event". Additionally, the "Authorization" header is included with a bearer token for authentication.

Note that the payload URL for the webhook should be set to the URL for the repository's dispatch endpoint with the appropriate values for <OWNER> and <REPO>.

> I am not sure if this can help, since I have never really tried it.

**Reply by mbhave** [reply]
No worries @Anmol-Baranwal and thanks for your reply. There isn't a way to customize the payload (at least I didn't find one). I ended up using an intermediate endpoint that the webhook can POST to that then triggers a `repository_dispatch`.

Since the solution is working for us now, I'm going to go ahead and close the discussion.

**Reply by pearljago** [reply]
Hi @mbhave I am really interested on how you finally manage to trigger the repository workflow by sending an organization event. As you mention early there is no chance to setup or edit a payload at organization level.

Thanks!

**Reply by wrzolekLukasz** [reply]
Hi all, 
I am facing similar issue and it seems that there is simply no way to add custom payload option to the webhook nor GitHuba Apps. The only one solution is to get request through middleware which is very last option for me. Is there anyone who figured out how to add custom payload on Git? 
@Anmol-Baranwal - would you be able to provide Screensoots where are you adding the payloads?

**Reply by Anmol-Baranwal** [reply]
Hi @wrzolekLukasz,
Please consult to the community. It has been some time since I did this.

**Reply by mbhave** [reply]
@pearljago I didn't find a way to do it directly so I wrote a layer that handles the webhook and triggers Github actions with the required `event_type`.

**Reply by anusreelakshmi934** [reply]
@pearljago Could you provide us with steps on how you did that? Thanks!
